teps/0049-aggregate-status-of-dag-tasks.md

```diff
@@ -0,0 +1,125 @@
+---
+status: proposed
+title: Aggregate Status of DAG Tasks
+creation-date: '2021-03-04'
+last-updated: '2021-03-08'
+authors:
+- '@pritidesai'
+---
+
+# TEP-0049: Aggregate Status of DAG Tasks
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+- [Proposal](#proposal)
+- [Test Plan](#test-plan)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+**Note** In this TEP, `dag` task refers to `pipelineTask` in the `tasks` section and `finally` task refers to
+`pipelineTask` in the `finally` section.
+
+A `finally` task in a `pipeline` has access to the execution status of any `dag` task through a context variable
+`$(tasks.<pipelineTask>.status)`. [TEP-0028: Task Execution Status at Runtime](0028-task-execution-status-at-runtime.md)
+implemented this context variable which is instantiated and available at the runtime. This context variable is set
+to `Succeeded`, `Failed`, or `None` depending on the execution status of that task and generally consumed in
+`when` expressions:
+
+```yaml
+    finally:
+      - name: notify-build-failure # notify build failure only when golang-build task fails
+        when:
+          - input: $(tasks.golang-build.status)
+            operator: in
+            values: ["Failed"]
+        taskRef:
+          name: send-to-slack-channel
+```
+
+
+This works great for a pipeline where a `finally` task depends on the execution status of an individual `dag` task.
+We often run into the use cases where a `finally` task needs to be executed if any one of the `dag` tasks fail.
+This use case can also be accomplished by explicitly spelling out each `dag` task in the comparison:
+
+```yaml
+finally:
+      - name: notify-any-failure # executed only when any dag task fail
+        when:
+          - input: "Failed"
+            operator: in
+            values: [$(tasks.unit-tests.status), $(tasks.golang-build.status), $(tasks.deploy.status)]
+```
+
+But, this solution does not scale well with a larger pipeline with 50 to 60 `dag` tasks. It explodes the `when`
+expressions with the entire list of `dag` tasks. Also, every time a pipeline is updated to add/delete a `dag` task,
+the pipeline author has to remember to update `when` expressions in the `finally` task as well.
+
+These concerns can be addressed by introducing a context variable with an aggregate status of all the `dag` tasks,
+`$(tasks.status)`.
+
+
+## Motivation
+
+* A `pipeline` author wants to perform some operation or execute a `finally` task as part of the same `pipeline` if one
+  of the `dag` tasks fail.
+
+### Goals
+
+* The main goal of this proposal is to allow access to aggregate status of all `dag` tasks within the `pipeline` at
```

Comment:
My understanding is that only finally tasks will have access to the new context variable. If so maybe make it more explicit here.

Reply:
thanks @souleb have made it explicit here and in the proposal section.